geode-redis/src/test/java/org/apache/geode/redis/internal/data/RedisKeyJUnitTest.java

```diff
@@ -0,0 +1,97 @@
+ * copyright ownership. The ASF licenses this file to You under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance with the License. You may obtain a
+ * copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License
+ * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+ * or implied. See the License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package org.apache.geode.redis.internal.data;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+import org.junit.BeforeClass;
+import org.junit.Test;
+
+import org.apache.geode.DataSerializer;
+import org.apache.geode.internal.HeapDataOutputStream;
+import org.apache.geode.internal.InternalDataSerializer;
+import org.apache.geode.internal.serialization.ByteArrayDataInput;
+import org.apache.geode.internal.serialization.DataSerializableFixedID;
+import org.apache.geode.redis.internal.executor.cluster.CRC16;
+
+public class RedisKeyJUnitTest {
+
+  @BeforeClass
+  public static void classSetup() {
+    InternalDataSerializer.getDSFIDSerializer()
+        .registerDSFID(DataSerializableFixedID.REDIS_KEY, RedisKey.class);
+  }
+
+  @Test
+  public void testRoutingId_withHashtags() {
+    RedisKey key = new RedisKey("name{user1000}".getBytes());
+    assertThat(key.getCrc16()).isEqualTo(CRC16.calculate("user1000"));
+
+    key = new RedisKey("{user1000".getBytes());
+    assertThat(key.getCrc16()).isEqualTo(CRC16.calculate("{user1000"));
+
+    key = new RedisKey("}user1000{".getBytes());
+    assertThat(key.getCrc16()).isEqualTo(CRC16.calculate("}user1000{"));
+
+    key = new RedisKey("user{}1000".getBytes());
+    assertThat(key.getCrc16()).isEqualTo(CRC16.calculate("user{}1000"));
+
+    key = new RedisKey("user}{1000".getBytes());
+    assertThat(key.getCrc16()).isEqualTo(CRC16.calculate("user}{1000"));
+
+    key = new RedisKey("{user1000}}bar".getBytes());
+    assertThat(key.getCrc16()).isEqualTo(CRC16.calculate("user1000"));
+
+    key = new RedisKey("foo{user1000}{bar}".getBytes());
+    assertThat(key.getCrc16()).isEqualTo(CRC16.calculate("user1000"));
+
+    key = new RedisKey("foo{}{user1000}".getBytes());
+    assertThat(key.getCrc16()).isEqualTo(CRC16.calculate("foo{}{user1000}"));
+
+    key = new RedisKey("{}{user1000}".getBytes());
+    assertThat(key.getCrc16()).isEqualTo(CRC16.calculate("{}{user1000}"));
+
+    key = new RedisKey("foo{{user1000}}bar".getBytes());
+    assertThat(key.getCrc16()).isEqualTo(CRC16.calculate("{user1000"));
+  }
+
+  @Test
+  public void testSerialization_withPositiveSignedShortCRC16() throws Exception {
+    RedisKey keyOut = new RedisKey("012345".getBytes());
+    assertThat((short) keyOut.getCrc16()).isPositive();
```

Comment:
It's not clear why this CRC value should be positive. Since we are trying to verify that the CRC16-XMODEM was correctly computed, this should test:

```
assertThat((short) keyOut.getCrc16()).isEqualTo((short)31712);
```

Reply:
The CRC16 value is really an unsigned short. Since shorts in Java are signed, we're holding it in an int. However, when serializing on the the wire, it is sent as a short to save some bits. When deserializing we `& 0xffff` the short to get it back to an `int` and effectively an unsigned short value. I wanted to make sure that the `& 0xffff` bit worked and the semantics of that were codified in a test.

Reply:
Oh I see @jdeppe-pivotal. I looked at that first assertion and stopped reading.